### DISCUSSION TITLE
filer内存占用问题请教

### DISCUSSION THREAD

**Discussion by paochiang** [seaweedfs/seaweedfs#3073] [Q&A]
在seaweedfs工作期间，观察到有时候filer的内存会增加比较大，直觉上我认为filer不操作数据，只操作元数据，感觉不应该用这么多的内存。（内存增加期间可能的相关数据操作是：多次修改同一文件）我想确认是什么地方导致filer使用的内存增加，但没有排查思路，想请教下能否提供排查思路和可能导致filer内存增加的代码模块？

版本：2.85

<img width="1766" alt="image" src="https://user-images.githubusercontent.com/28077875/168991259-8347f2e2-addd-4e9e-9a84-c9f6ac9f2343.png">

**Comment by paochiang** [comment]
我在另一个环境复现了一下这个问题，同时filer开了debug模式：
prometheus监控：
<img width="1700" alt="image" src="https://user-images.githubusercontent.com/28077875/169446215-c30f4425-041d-4f65-805d-33f7155936f2.png">

VmRSS:
<img width="1028" alt="image" src="https://user-images.githubusercontent.com/28077875/169446121-87cd55e4-0be0-412a-9fd9-46fca4ae6802.png">

pprof inuse_space:
<img width="1699" alt="image" src="https://user-images.githubusercontent.com/28077875/169446027-081bc530-5b45-4cb7-a8ea-763b398a8b01.png">

pprof文件(在Prometheus监控图最右侧时间点保存pprof文件，可能有几分钟的时间差距)：[pprof.weed.alloc_objects.alloc_space.inuse_objects.inuse_space.012.pb.gz](https://github.com/chrislusf/seaweedfs/files/8735770/pprof.weed.alloc_objects.alloc_space.inuse_objects.inuse_space.012.pb.gz)

pprof采样到的内存使用和VmRSS获取到的内存使用差别很大，我没能从inuse_spaces上看出明显问题。

通过inuse_objects数据，我看着可能和下面代码有关，但不是很确定：
https://github.com/chrislusf/seaweedfs/blob/3a8dadcc2d5bfc5edb63395bbfb4f502c8333114/weed/util/log_buffer/log_read.go#L41

补充实时的pprof（时间可能和上面差了十分钟左右）:
<img width="365" alt="image" src="https://user-images.githubusercontent.com/28077875/169448178-6a5b0d97-580f-442e-8cb6-c4a7893fce55.png">

<img width="1688" alt="image" src="https://user-images.githubusercontent.com/28077875/169448059-b57d4f69-9674-4ee7-95bb-c9c17f01d1da.png">

**Reply by chrislusf** [reply]
Thanks for the investigation! The memory seems taken by the metadata logs. How busy is the write operation?

**Reply by paochiang** [reply]
<img width="1689" alt="image" src="https://user-images.githubusercontent.com/28077875/169463620-a17f3fa0-d951-4d57-b0e5-3cf4e8885783.png">
主要写操作是修改同一个文件

**Reply by chrislusf** [reply]
how large is the file, how large is each write, are they appending operations?

**Reply by paochiang** [reply]
最开始大小为0，从2022-05-19 15:55(CST)写到 2022-05-20 14:52，一共追加写到了1.8MB
<img width="806" alt="image" src="https://user-images.githubusercontent.com/28077875/169470188-6e463586-138e-4a1b-a32f-29b198ae4524.png">

每次写入的内容很少，但具体是多少不清楚。（因为这个读写操作是跑一个AI训练任务过程中发生的，训练任务要跑好几天）

我在另一个环境，开过-v 4日志发现整个过程中都是append操作，其中的一次操作日志如下：
```
I0516 19:38:17 27131 file.go:51] file Attr /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0, open:0 existing:valid=1m0s ino=0 size=0 mode=----------
I0516 19:38:17 27131 file.go:98] file /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 open Open [ID=0x3db Node=0x5 Uid=0 Gid=0 Pid=32265] dir=false fl=OpenWriteOnly+OpenAppend
I0516 19:38:17 27131 wfs.go:154] AcquireHandle /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 uid=0 gid=0
I0516 19:38:17 27131 wfs.go:178] Acquired new Handle /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 open 1
I0516 19:38:17 27131 file.go:105] /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 file open handle id = 871604664355894590
I0516 19:38:17 27131 page_writer.go:38] /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 AddPage [4754, 4794) streaming:false
I0516 19:38:17 27131 dirty_pages_temp_file.go:40] /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 tempfile AddPage [4754, 4794)
I0516 19:38:17 27131 filehandle.go:246] Flush /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 fh 871604664355894590
I0516 19:38:17 27131 filehandle.go:270] doFlush /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 fh 871604664355894590
I0516 19:38:17 27131 file.go:329] /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 existing 113 chunks adds 1 more
I0516 19:38:17 27131 dirty_pages_temp_file.go:93] /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 saveToStorage 2,4a90bfb97b47 [4754,4794)
I0516 19:38:17 27131 filehandle.go:310] /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 set chunks: 114
I0516 19:38:17 27131 filehandle.go:312] /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 chunks 0: 1,4a11475bf06d [0,40)
I0516 19:38:17 27131 filehandle.go:312] /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 chunks 1: 1,4a1488254abf [40,84)
.....
I0516 19:38:17 27131 filehandle.go:312] /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 chunks 104: 1,4a87dd48af97 [4374,4414)
I0516 19:38:17 27131 filehandle.go:312] /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 chunks 105: 2,4a8840a36aa0 [4414,4455)
I0516 19:38:17 27131 filehandle.go:312] /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 chunks 106: 1,4a89d281d68d [4455,4496)
I0516 19:38:17 27131 filehandle.go:312] /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 chunks 107: 3,4a8a85767af7 [4496,4538)
I0516 19:38:17 27131 filehandle.go:312] /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 chunks 108: 3,4a8bf1ceb620 [4538,4582)
I0516 19:38:17 27131 filehandle.go:312] /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 chunks 109: 1,4a8c81b14024 [4582,4628)
I0516 19:38:17 27131 filehandle.go:312] /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 chunks 110: 1,4a8df0eec367 [4628,4670)
I0516 19:38:17 27131 filehandle.go:312] /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 chunks 111: 3,4a8eebfae399 [4670,4714)
I0516 19:38:17 27131 filehandle.go:312] /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 chunks 112: 2,4a8f5e7a9c7e [4714,4754)
I0516 19:38:17 27131 filehandle.go:312] /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 chunks 113: 2,4a90bfb97b47 [4754,4794)
    I0516 19:38:17 27131 filerstore_wrapper.go:128] InsertEntry /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0
I0516 19:38:17 27131 filehandle.go:218] Release /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 fh 871604664355894590 open=1
I0516 19:38:17 27131 wfs.go:186] ReleaseHandle /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0 id 871604664355894590 current handles length 1
I0516 19:38:17 27131 file.go:51] file Attr /lbj/exp0/runs/events.out.tfevents.1652700959.dev21.334.0, open:0 existing:valid=1m0s ino=0 size=0 mode=
```
可以看到文件的打开方式是：`OpenWriteOnly+OpenAppend`

当时-v日志大概收集了十多个小时我就停止了任务运行，一共18GB的日志数据，

<img width="1187" alt="image" src="https://user-images.githubusercontent.com/28077875/169472436-aa9c1945-bccb-4a26-8462-903734905ea9.png">

**Reply by paochiang** [reply]
AI任务训练过程中，也会读和写其他seaweedfs中的文件。我做过测试，只要不将上面的这个文件`events.out.tfevents.1652946955.aca987bd5f24e7611e8985ea3d9d4737-taskrole1-xxxx`写入到seaweedfs存储，filer的内存会很稳定，不会增加，所以大概能确定是append这个文件导致的filer内存上升。

但奇怪的是我自己尝试用golang append文件到seaweedfs（append 1000次），却复现不出filer内存增加的情况。我目前只能用特定的AI训练任务复现出这个情况。
